ar71xx: add support for TP-Link CPE210 v2

robimarko opened the pull request:
This PR adds support for a popular low-cost 2.4GHz N based AP

Specifications:
 - SoC: Qualcomm Atheros QCA9533 (650MHz)
 - RAM: 64MB
 - Storage: 8 MB SPI NOR
 - Wireless: 2.4GHz N based built into SoC 2x2
 - Ethernet: 1x 100/10 Mbps, integrated into SoC, 24V POE IN

Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Notes:
TP-Link does not use bootstrap registers so without this patch reference
clock detects as 40MHz while it is actually 25MHz.
This is due to messed up bootstrap resistor configuration on the PCB.
Provided GPL code just forces 25MHz reference clock.
That causes booting with completely wrong clocks, for example, CPU tries
to boot at 1040MHz while the stock is 650MHz.
So this PR depends on PR #672 to remove 40MHz reference clock.
Thanks to Sven Eckelmann <sven@narfation.org> for properly patching that.

Signed-off-by: Robert Marko <robimarko@gmail.com>

mkresin commented:
Please, open a PR in the [OpenWrt repository](https://github.com/openwrt/openwrt/pulls).

With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed during the next days.

mkresin commented:
Sorry, wrong tab.

blogic commented on target/linux/ar71xx/patches-4.4/999-qca953x-clock-25mhz.patch:
```diff
@@ -0,0 +1,62 @@
++	} else
+ 		ref_rate = 25 * 1000 * 1000;
+ 
+ 	pll = ath79_pll_rr(QCA953X_PLL_CPU_CONFIG_REG);
```
might be cleaner code to write 

if (mips_machtype == ATH79_MACH_CPE210_V2)
...
else if (bootstrap &	QCA953X_BOOTSTRAP_REF_CLK_40) 
...
else
...

blogic commented on target/linux/ar71xx/patches-4.4/999-qca953x-clock-25mhz.patch:
```diff
@@ -0,0 +1,62 @@
++		else
++			ath79_wmac_data.is_clk_25mhz = false;
++	} else
+ 		ath79_wmac_data.is_clk_25mhz = true;
```
same here, less indentation is often easier to read.

blogic commented on target/linux/ar71xx/patches-4.9/999-qca953x-clock-25mhz.patch:
```diff
@@ -0,0 +1,54 @@
++		else
++			ref_rate = 40 * 1000 * 1000;
++	} else
+ 		ref_rate = 25 * 1000 * 1000;
```
and here

blogic commented on target/linux/ar71xx/patches-4.9/999-qca953x-clock-25mhz.patch:
```diff
@@ -0,0 +1,54 @@
++			ath79_wmac_data.is_clk_25mhz = true;
++		else
++			ath79_wmac_data.is_clk_25mhz = false;
++	} else
```
and here...

ecsv commented:
Thanks for the feedback

@robimarko  The adjusted patches can be found in https://github.com/ecsv/openwrt/tree/CPE210-v2. I have also rebased it to avoid the conflict created by the move of target/linux/ar71xx/image/tp-link.mk to target/linux/ar71xx/image/generic-tp-link.mk

robimarko commented:
@ecsv Thanks again,I applied you changes on rebased branch.
All credits to you again.

robimarko commented:
@blogic Changes made

pepe2k commented on target/linux/ar71xx/patches-4.4/703-MIPS-ath79-fix-CPE210-v2-refclk.patch:
```diff
@@ -0,0 +1,60 @@
+ 
+ 	bootstrap = ath79_reset_rr(QCA953X_RESET_REG_BOOTSTRAP);
+-	if (bootstrap &	QCA953X_BOOTSTRAP_REF_CLK_40)
++	if (mips_machtype == ATH79_MACH_CPE210_V2)
```
Personally, I think that using here a statement based on mach type might make things more complicated in (near) future. I wouldn't be surprised to see more TP-Link boards coming with the same approach which would require updating this patch over and over again.

Based on all QCA953x datasheets which [are available on-line](https://github.com/Deoptim/atheros/), this SoC series just doesn't support other reference clock type than 25 MHz and I haven't seen board with different clock so far.

![qca953x-bootstrap](https://user-images.githubusercontent.com/2934612/35148198-a5e6ab52-fd11-11e7-9f5d-dbe4068a718c.png)

I'm not going to block this PR because of these two patches but I think that it would be safe to just set a static 25 MHz reference clock for QCA953x SoC series.

@ecsv, as you are the author of the patches, what is your opinion?

pepe2k reviewed (changes requested):
Hi @robimarko,

Thank you for your PR, please see my comments inline, in code. Also, please wrap your commit description (according to our [submitting patches guide](https://lede-project.org/submitting-patches)) and fix some typos (`bootsrap`).

Cheers,
Piotr

ecsv commented:
I am fine with it. I am unable to make sure that there are no boards which actually need 40Mhz support but I can make different patch to remove 40MHz completely for QCA953x. (But I have to finish some other things first before I can do that.... so feel free to do it yourself when you are in a hurry)

pepe2k commented on target/linux/ar71xx/base-files/etc/board.d/01_leds:
```diff
@@ -213,16 +213,28 @@ cf-e530n)
+	cpe210|\
+	cpe510|\
+	wbs210|\
+	wbs510))
```
This is a bug.
And, why not use `cpe210-v2)` and `*)`?

robimarko commented on target/linux/ar71xx/base-files/lib/ar71xx.sh:
```diff
@@ -361,6 +361,25 @@ tplink_pharos_board_detect() {
 	AR71XX_MODEL="TP-Link $model v$2"
 }
 
+tplink_pharos_v2_get_model_string() {
```
Well, TP-Link pretty much completely changed product-info partition.
So I haven't thought of a good way to do the checks in existing functions.

robimarko committed:
ar71xx: Add TP-Link Pharos v2 board detection

Add support for detecting TP-Link Pharos v2 boards.
They use different format in product-info partition than v1 boards.

Code for this patch was written by Alexander Couzens <lynxis@fe80.eu>

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
ar71xx: add support for TP-Link CPE210 v2

This PR adds support for a popular low-cost 2.4GHz N based AP

Specifications:
 - SoC: Qualcomm Atheros QCA9533 (650MHz)
 - RAM: 64MB
 - Storage: 8 MB SPI NOR
 - Wireless: 2.4GHz N based built into SoC 2x2
 - Ethernet: 1x 100/10 Mbps, integrated into SoC, 24V POE IN

Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Notes:
TP-Link does not use bootstrap registers so without this patch reference
clock detects as 40MHz while it is actually 25MHz.
This is due to messed up bootstrap resistor configuration on the PCB.
Provided GPL code just forces 25MHz reference clock.
That causes booting with completely wrong clocks, for example, CPU tries
to boot at 1040MHz while the stock is 650MHz.
So this PR depends on PR #672 to remove 40MHz reference clock.
Thanks to Sven Eckelmann <sven@narfation.org> for properly patching that.

Signed-off-by: Robert Marko <robimarko@gmail.com>

ecsv commented:
The non-40Mhz patch for QCA953x was added last Thursday to OpenWrt. Maybe a good idea to rebase your patch and check that it still works.

blocktrron commented:
@robimarko @ecsv  I rebased the branch and flashed it on my CPE210 v2. Everything works as expected, also the WiFi MAC-address matches the factory one.

In case it matters, my rebased branch is here: https://github.com/blocktrron/openwrt/tree/CPE210-v2

robimarko commented:
Thanks for checking.
Will rebase it and push soon.
I still gotta reduce code duplication

mweinelt commented:
@robimarko Not to be pushy, but do you still have this on your todo list?

robimarko commented:
Yes definitively

adrianschmutzler commented:
Watch out: In commits 373dc11b519516682ab48bf7db501ffffced4e61 and e39847ea2f70bf8880a4533b22a0eed110ffea98 the kernel partition size for the cpe210/510 v1 was changed. The device definition in generic-tp-link.mk automatically inherits those changes while the partitions in tplink-safeloader.c remain unchanged.

Result: bootloop!

This won't be a problem for backports to 17.01 (gluon), but will break builds based on the OpenWRT master. Should be fixable easily be doing the same adjustments in tplink-safeloader.c as for the cpe210v1.

robimarko commented:
Will update the PR this week to fix the issue,so far it has been a really busy week

adrianschmutzler commented:
Adjusted the partitions in the last commit here (the patches before are 1:1 copies from blocktrron) :
https://github.com/adrianschmutzler/openwrt/commits/cpev2_2r

Technically, one should squash this into commit 2. However, for backporters a separate patch for the partition size changes would be very helpful.

rotanid commented:
this is obsoleted by #937 i guess
